src/sentry/api/endpoints/organization_member_details.py

```diff
@@ -129,22 +129,20 @@ def put(self, request, organization, member_id):
         if not serializer.is_valid():
             return Response(status=400)
 
-        has_sso = AuthProvider.objects.filter(
-            organization=organization,
-        ).exists()
+        auth_provider = AuthProvider.objects.filter(organization=organization)
```

Comment:
Don't you want this to be a `.get` and not `.filter`? Otherwise you're returning a `QuerySet` that hasn't been evaluated yet.

Reply:
Actually I think I'll go with `.first`, since that will return None in that case that there isn't one.

So I won't have to

```py
try:
    auth_provider = AuthProvider.objects.get(organization=organization)
except AuthProvider.DoesNotExist:
    auth_provider = None
```

But yeah, I was passing a query set into the send_sso_link_email as the provider, and the email happend to just read correctly even though the provider.name was missing ;-)

> Single Sign-On has been configured for the organization. Link your Sentry account to enable signing on with your account.

Reply:
So, I'd like to point this out, because for some reason people keep wanting to use `.first()` I think without realizing the significant differences. It doesn't quite mean the same thing, and I'd prefer we use it correctly instead of using it for syntax sugar.

`.first()` applies an ordering to the queryset. By default, this would be adding an `ORDER BY id LIMIT 1` since `first` is implying an ordered set. Some times this `ORDER BY` will change the query plan resulting in an unoptimized query.

Now, given that, `.get()` doesn't add a `LIMIT` or `ORDER BY`. The behavior of `get()` is to potentially raise two different exceptions. `ObjectDoesNotExist` and `MultipleObjectsReturned`. So this will both error when there are no rows, and when more than 1 row has been returned. This is ideal when you are working with something that already has a `UNIQUE` and you do want to explicitly error if there were more than 1 row matched. Potentially if the result were ambiguous. I'd say _this_ behavior is correct here since there shouldn't be more than 1 `AuthProvider` matched. And if there were more than 1 row, I think we'd want to error since linking auth would be ambiguous if more than 1 were returned.

The last real alternative is `[0]` which is kind of in the middle. It results in a `LIMIT 1` added to the query, but raises an `IndexError` instead of `ObjectDoesNotExist`, and will not error if there are more than 1 row, because it doesn't know since there was a database level `LIMIT` applied. But unlike `first()` this does not introduce an `ORDER BY`.

Given this information, I'd really suggest we just use the normal `try..except` pattern here since it's mostly what we want without introducing different behavior because of `.first()`. Personally, I'd prefer we didn't use `first()` at all, since it seems it's purely used for syntax sugar without realizing it changes semantic meaning of the underlying query.

We talked about just maybe adding a simple helper like, `QuerySet.fetch()` or something that would just be the syntax sugar for the `try...except` pattern that people seem to hate, which would just return `None`.

Reply:
Thank you for clarifying @mattrobenolt, I wasn't aware that it stuck an `ORDER` on there.

> since it seems it's purely used for syntax sugar without realizing it changes semantic meaning of the underlying query.

Agreed with this, and that generally does become the burden of abstractions like `QuerySet`.

I will switch this over to a `try` `except` get.